Do not restart browser when there is test type switch

WeizhongX opened the pull request:
Do not restart browser when test type changes

At least for Chrome, parameters used to start Chrome won't change when
test type changes.

This change tries to switch to a different executor where allowed, and
reuse the webdriver session from the previous executor.

A browser side flag is used to control whether this behavior should be
enabled or not. It is up to each browser vendor to decide whether it
should be enabled or not.

Bug: 341581718

jgraham commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -558,6 +569,16 @@ def start_test_runner(self):
         # Now we wait for either an init_succeeded event or an init_failed event
 
+
+    def get_executor_for_test_type(self, test_type):
```
I don't like this being different to the way we handle setting the initial executor. I'm also nervous that it depends on the class instance being serializable cross-process. I think maybe `start_runner` and `set_executor` should take `(executor_cls, executor_kwargs, executor_browser_cls, executor_browser_kwargs)` (maybe wrapped in a named tuple) and the `TestRunner` class should have a `set_executor` method that's used from `__init__` and when passed the `set_executor` message that contructs the executor.

WeizhongX replied:
Thanks, updated but not exactly same as what you suggested, because `set_executor` needs to do something additional, i.e. tear down previous executor and setup the new one.

PTAL

WeizhongX commented on tools/wptrunner/wptrunner/browsers/base.py:
```diff
@@ -103,6 +103,10 @@ def setup(self):
         """Used for browser-specific setup that happens at the start of a test run"""
         pass
 
+    def needs_restart(self, new_test_type, old_test_type):
```
Done, thanks!

WeizhongX commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -78,6 +85,16 @@ def setup(self):
+            **executor_implementation.executor_browser_kwargs)
+        self.executor = executor_implementation.executor_cls(
+            browser, **executor_implementation.executor_kwargs)
+        self.executor.setup(self)
```
Recreate `TestRunner` can have a similar problem that executor setup can fail. This can be solved if we can have a new state in the main process.

WeizhongX replied:
Hi @gsnedders, want to get your opinion/suggestion on this. Thanks!

gsnedders replied:
I think @jgraham's suggestion of just recreating the `TestRunner` would lead to a smaller PR?

From a purity point-of-view, it's not clear to me which approach is actually better. There's certainly some simplicity afforded by `TestRunner` only dealing with a single executor.

We do currently fail if the executor fails to connect to the browser, I thought? Which implies we do handle construction the executor today?

WeizhongX replied:
Currently creating TestRunner is part of the init process, so if we recreate TestRunner, it needs to be out of that process, and implies a new state if we still want to handle executor setup failures. I will make a change for review if you do not object.

construction of the executor is handled here:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/testrunner.py;drc=9bf3c522b764982e35548f83f3253741f70454f7;l=158

WeizhongX replied:
I plan to restart this work.

> There's certainly some simplicity afforded by TestRunner only dealing with a single executor.
I agree with this. Maybe the Executor should be able to handle test types by itself, instead of having different Executors for different test type. But the change would be too big.

jonathan-j-lee replied:
Reusing the protocol (i.e., WebDriver session) seems to require a lot of plumbing:
* A `TestRunnerManager` `switching` state
* A `switch_executor` `TestRunner` command
* A `setup(..., protocol)` arg (also, `setup`/`teardown` not symmetric; see other comment)

What is the overhead of creating a new WebDriver session relative to restarting `chromedriver`? If it's not too high, maybe we can just throw away the executor process, reusing the existing `chromedriver`, and not try to work against testrunner's existing fault lines. Unfortunately, I imagine most of the performance gain comes from reusing a session, since each is a different `chrome`.

WeizhongX commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -261,10 +261,10 @@ class TestExecutor:
 
 
-    def __init__(self, logger, browser, server_config, timeout_multiplier=1,
+    def __init__(self, browser, server_config, timeout_multiplier=1,
```
Thanks! Previously in this PR executor class is constructed outside of TestRunner, so to make sure executor always get same logger as TestRunner, removed the logger from here, and set up the logger during executor setup.

Now this is not needed as we are constructing executor inside TestRunner. Will revert this part of change.

WeizhongX committed:
Do not restart browser when test type changes

At least for Chrome, parameters used to start Chrome won't change when
test type changes (except for wdspec tests).

This change tries to switch to a different executor where allowed, and
reuse the webdriver session from the previous executor.

A browser side flag is used to control whether this behavior should be
enabled or not. It is up to each browser vendor to decide whether it
should be enabled or not.

Bug: 341581718

WeizhongX committed:
add type annotation

WeizhongX commented:
@jonathan-j-lee pls help take a look. The try run is here: https://ci.chromium.org/ui/p/chromium/builders/try/linux-wpt-chromium-rel/1415/overview.

WeizhongX committed:
fix unit tests

WeizhongX commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -108,6 +115,22 @@ def teardown(self):
+        protocol = self.executor.protocol
+        browser = executor_implementation.executor_browser_cls(
+            **executor_implementation.executor_browser_kwargs)
+        self.executor = executor_implementation.executor_cls(
```
`teardown` will be called eventually when the connection is not needed. This is not a problem for chrome's executors, but some other executors are doing additional things in `setup`, so this function can not be applied directly for them.

WeizhongX replied:
This can be improved. I actually should set `protocol` to None and still have `teardown` called for the previous executor.

jonathan-j-lee replied:
> set protocol to None and still have teardown called for the previous executor.

SGTM

WeizhongX commented on tools/wptrunner/wptrunner/browsers/chrome.py:
```diff
@@ -194,6 +194,12 @@ def update_properties():
 
 class ChromeBrowser(WebDriverBrowser):
+    def restart_on_test_type_change(self, new_test_type: str, old_test_type: str) -> bool:
+        # restart the test runner when switch from/to wdspec tests
+        if "wdspec" in [old_test_type, new_test_type]:
+            return True
+        return False
```
This is a good idea.

WeizhongX commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -108,6 +115,22 @@ def teardown(self):
         self.browser = None
 
+    def switch_executor(self, executor_implementation):
+        assert self.executor is not None
```
True, and that is the reason why `assert` is used instead of an if statement, right? Or did I miss your point?

A similar use of `assert` below:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/testrunner.py;l=555

jonathan-j-lee replied:
Sorry, what I meant was that mypy (which this repo has) can verify the non-nullibility of `TestRunner.executor` statically, whereas the `TestRunnerManager` state can't be.

WeizhongX commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -301,6 +326,8 @@ class _RunnerManagerState:
     running = namedtuple("running", ["subsuite", "test_type", "test", "test_group", "group_metadata"])
     restarting = namedtuple("restarting", ["subsuite", "test_type", "test", "test_group",
                                            "group_metadata", "force_stop"])
+    switching = namedtuple("switching", ["subsuite", "test_type", "test", "test_group",
```
This is something I am debating myself also. Sure let's do that.

WeizhongX committed:
update per review comments

jonathan-j-lee commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -844,12 +900,20 @@ def after_test_end(self, test, restart, force_rerun=False, force_stop=False):
+                    restart = True
+                else:
+                    self.logger.info(f"Switching executor for new test type: {self.state.test_type} => {test_type}")
+                    impl = self.test_implementations[(subsuite, test_type)]
```
```suggestion
                    impl = self.test_implementations[subsuite, test_type]
```

jonathan-j-lee commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -93,7 +100,7 @@ def __exit__(self, exc_type, exc_value, traceback):
         self.logger.debug("Executor setup")
         try:
-            self.executor.setup(self)
+            self.executor.setup(self, None)
```
```suggestion
            self.executor.setup(self)
```

jonathan-j-lee commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -630,9 +633,9 @@ def __init__(self, logger, browser, server_config, webdriver_binary,
+    def setup(self, runner, protocol=None):
         self.protocol = self.protocol_cls(self, self.browser)
-        super().setup(runner)
+        super().setup(runner, None)
```
```suggestion
        super().setup(runner)
```

WeizhongX committed:
Apply suggestions from code review

Co-authored-by: Jonathan Lee <jonathan-j-lee@users.noreply.github.com>

jgraham commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -844,12 +900,20 @@ def after_test_end(self, test, restart, force_rerun=False, force_stop=False):
                 self.logger.info("Restarting browser for new test group")
                 restart = True
+            elif test_type != self.state.test_type:
+                if self.browser.browser.restart_on_test_type_change(test_type, self.state.test_type):
```
Since there's also functional requirements that the `browser` class doesn't know about (i.e. whether the two test types share a protocol) we should actually have a method in this class that does a pre-check before deferring to the browser class iff the pre-check passes.

WeizhongX replied:
jonathanjlee@ made a similar suggestion previously: https://github.com/web-platform-tests/wpt/pull/40556#discussion_r1608698082.

Looks like this is kind of the same as your comment in `tools/wptrunner/wptrunner/executors/base.py`. How about we check protocol type in `switch_executor`, and if protocol type does not match, we would let `switch_executor` to fail?

WeizhongX committed:
check protocol class when switch executor

WeizhongX commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -286,13 +286,16 @@ def __init__(self, logger, browser, server_config, timeout_multiplier=1,
+        :param protocol: protocol connection to reuse if not None"""
         self.runner = runner
-        if self.protocol is not None:
+        if protocol is not None:
```
Added the check, thanks!

WeizhongX committed:
assert protocol class matches in setup

WeizhongX commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -630,7 +633,7 @@ def __init__(self, logger, browser, server_config, webdriver_binary,
         self.binary_args = binary_args
 
-    def setup(self, runner):
+    def setup(self, runner, protocol=None):
```
Done, thanks!

WeizhongX commented:
The failure on infrastructure/tests is not related with this change.

@KyleJu @jcscottiii can you help admin merge this?

This has been open for too long, and as there is no response from jgraham@ after multiple pings, let's merge this for now. I think his comments have been addressed.
